Sources/NIOHPACK/HPACKEncoder.swift

```diff
@@ -0,0 +1,249 @@
+    public var encodedData: ByteBuffer {
+        if let newSize = self.newDynamicTableSize {
+            // prepend an encoded table resize command
+            var result = self.allocator.buffer(capacity: self.buffer.readableBytes + 11)    // largest possible required space
+            result.write(encodedInteger: UInt(newSize), prefix: 5, prefixBits: 0x20)
+            result.write(bytes: self.buffer.readableBytesView)  // write(buffer:) will remove bytes read from the input buffer
+            return result
+        } else {
+            return self.buffer
+        }
+    }
+    
+    /// Whether to use Huffman encoding.
+    public let useHuffmanEncoding: Bool
+    
+    /// The current size of the dynamic table.
+    ///
+    /// This is defined as the sum of [name] + [value] + 32 for each header.
+    public var dynamicTableSize: Int {
+        return self.headerIndexTable.dynamicTableLength
+    }
+    
+    /// The current maximum size to which the dynamic header table may grow.
+    public private(set) var allowedDynamicTableSize: Int {
+        get { return self.headerIndexTable.dynamicTableAllowedLength }
+        set { self.headerIndexTable.dynamicTableAllowedLength = newValue }
+    }
+    
+    /// The hard maximum size of the dynamic header table, set via an HTTP/2
+    /// SETTINGS frame.
+    public var maximumDynamicTableSize: Int {
+        get { return self.headerIndexTable.maxDynamicTableLength }
+        set { self.headerIndexTable.maxDynamicTableLength = newValue }
+    }
+    
+    /// Sets the maximum size for the dynamic table and encodes the new value
+    /// at the start of the current packed header block to send to the peer.
+    ///
+    /// - note: The size cannot grow beyond the maximum allotted by the most recent
+    ///         acknowledged SETTINGS_HEADER_TABLE_SIZE. It will truncate to that
+    ///         value automatically.
+    ///
+    /// - Parameter size: The new maximum size for the dynamic header table.
+    public mutating func setDynamicTableSize(_ size: Int) {
+        // this can only be encoded at the beginning of a header block. Therefore, we keep it
+        // on the side and prepend it in when handing out our encoded data.
+        if let existingNewSize = self.newDynamicTableSize {
+            // only change things when growing the buffer. Don't grow then shrink, because
+            // we will send only the larger size to the peer.
+            if size > existingNewSize && size <= self.headerIndexTable.maxDynamicTableLength {
+                let newSize = min(max(size, existingNewSize), self.headerIndexTable.maxDynamicTableLength)
+                self.newDynamicTableSize = newSize
+                self.allowedDynamicTableSize = newSize
+            }
+        } else {
+            let newSize = min(size, self.headerIndexTable.maxDynamicTableLength)
+            self.newDynamicTableSize = newSize
+            self.allowedDynamicTableSize = newSize
+        }
+    }
```

Comment:
I'm pleased that we've added support for staying in the bounds of RFC 7541, but I think we may have overcomplicated this a bit, and we're also missing a behaviour that RFC 7541 requires of us. So let's step back a minute.

Firstly, the over complication. This code works pretty hard to keep track of the new max table size, which is good. However, because this encoder allows a "streaming" encode mode, it raises the possibility of users making this sequence of calls:

```swift
encoder.setDynamicTableSize(someVal)
encoder.append(header: "first", value: "first-val")
encoder.setDynamicTableSize(someOtherVal)
encoder.append(header: "second", value: "second-val")
encoder.setDynamicTableSize(someOtherValAgain)
encoder.append(header: "third", value: "third-val")
```

This call sequence is bad, as only the third value will be reflected in the encoded header block, but the encoder should have respected all the interim values and done the appropriate evictions if needed. (Incidentally, does this encoder actually evict from the table if we change the header table length? It doesn't appear to, but I may have missed it.)

Arguably the allowance of a streaming encode mode is the issue here, but if we are going to keep supporting it we need to be keeping track of state.

The second issue is this stanza of RFC 7541 § 4.2:

> Multiple updates to the maximum table size can occur between the transmission of two header blocks.  In the case that this size is changed more than once in this interval, the smallest maximum table size that occurs in that interval MUST be signaled in a dynamic table size update.  The final maximum size is always signaled, resulting in at most two dynamic table size updates.  This ensures that the decoder is able to perform eviction based on reductions in dynamic table size (see Section 4.3).

This means that we need to keep track of *two* numbers and a flag. The two numbers we need to keep track of are the current max table size and the smallest it's been since the last time we encoded. The flag we need is whether the max table size has changed.

Given that we *also* need some state to prevent the user from changing the max dynamic table size during encode of a header block, I think we need to refactor the API and internals of this object quite a bit.

The state management is probably best done with an enum:

```swift
private enum EncoderState {
    case idle(maxTableSize: Int)
    case resized(currentMaxTableSize: Int, smallestMaxTableSize: Int?)
    case encoding(maxTableSize: Int)
}
```

Once we move to this explicit state management, we can now have `reset` trigger the change to the `.encoding` state, and as a result it can automatically write out any table size changes that are needed. This avoids the copy we introduced in `encodedData`.

Speaking of `encodedData`, we can no longer use a property for this, as we need an explicit signal that we are done encoding a block. We should probably add a function called `finalize()` or something equivalent that completes a block and returns the buffer. That will transition us back to `idle`, ready to do the whole dance again.

How does this sound to you?